File: javascript/arrays-and-hashing/group-anagrams.js
var groupAnagrams = function (strs) {
    // to solve this problem in the easiest and most straightforward way make a map for of each string then use the 
    const anagrams = new Map()

    for (let str of strs) {
        let sorted = str.split('').sort().join('')

        if (!anagrams.has(sorted)) {
            anagrams.set(sorted, [])
        }
        anagrams.get(sorted).push(str)
    }
    
    return Array.from(anagrams.values())
};


console.log(groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]))
console.log(groupAnagrams([""]))
console.log(groupAnagrams(["a"]))

File: python/neetcode/arrays-and-hashing/group-anagrams.py
from collections import defaultdict

def groupAnagrams(strs):
    
    anagram_groups = defaultdict(list)

    for s in strs:
        sorted_str = ''.join(sorted(s))
        anagram_groups[sorted_str].append(s)
        # print(sorted_str)

    print(anagram_groups.values())

    return list(anagram_groups.values())

groupAnagrams(["eat","tea","tan","ate","nat","bat"])